### DISCUSSION TITLE
How to get the output of a docker exec command when running remote script on a Linux host over SSH

### DISCUSSION THREAD

**Discussion by olavaukan** [PowerShell/PowerShell#17573] [Q&A]
Apologies if this is the wrong place for this sort of question.

I have PowerShell remoting working over SSH from Windows 10 to Linux (CentOS 7), but I'm struggling with using this in combination with Docker and capturing/redirecting the output of a `docker exec` command to run a query inside a PostgreSQL container.

Basically I have PostgreSQL running inside a Docker container and not exposing its ports to the host, so I need to run the commands on the container itself or another container in the same network.

If I SSH into the server and run the command I get something like this, which is the expected output:

```
$ docker exec database bash -c 'psql -U dbUser -d dbName -c "SELECT COUNT(*) FROM public.\"Resources\" WHERE \"CountyId\" IS NULL"'
 count
-------
 77765
(1 row)
```

If I try the same thing remotely using Invoke-Command I seem to miss the "payload":

```
$Session = New-PSSession -HostName host -UserName user
Invoke-Command $session -ScriptBlock { docker exec database bash -c 'psql -U dbUser -d dbName -c "SELECT COUNT(*) FROM public.Resources WHERE CountyId IS NULL"' }
--
(1 row)
```

If I try adding the `-it` flag to `docker exec` I get an error:

```
$Session = New-PSSession -HostName host -UserName user
Invoke-Command $session -ScriptBlock { docker exec -it database bash -c 'psql -U dbUser -d dbName -c "SELECT COUNT(*) FROM public.Resources WHERE CountyId IS NULL"' }
the input device is not a TTY
```

I've tried setting `PermitTTY yes` in `/etc/ssh/sshd_config` on the host and `RequestTTY yes` in `~\.ssh\config` on the client but this results in another error:

```
$Session = New-PSSession -HostName host -UserName user
New-PSSession: [host] The background process reported an error with the following message: The SSH client session has ended with error message: Pseudo-terminal will not be allocated because stdin is not a terminal..
```

So to summarize I want to:
1. Use PowerShell Remoting to establish an SSH session from a Windows client to a Linux host.
2. Use `docker exec` to run a `bash` command on an existing PostgreSQL container on that Linux host.
3. Use that `bash` command to run a `psql` command.
4. Use that `psql` command to run a `SELECT` query on a database.
5. Return the result of that query back up through these layers all the way back to PowerShell.

I realize this is getting superconvoluted, but AFAIK this should work fine with standard SSH and running a bash-script remotely. For various reasons I'm trying to stick to a single scripting language for this project and PowerShell Core seemed like the best option, but I seem to have come up against some limitations here.

**Comment by kasini3000** [comment]
I think:
powershell 5 or 7 + .net postgresql client library is the authentic way to solve the problem. 
It is thread level, can return **sqlrow** object, can return **sqltable** object, can bypass your problem.

https://www.nuget.org/packages/Npgsql/7.0.0-preview.5#supportedframeworks-body-tab

You can use this method on the master computer.
or through the powershell batch scripting framework kasini3000 written by me, on the node computer.

eg:
on master:

```
$sql_table_object_from_node_to_master = k_run_ip -ip nodeip { pg_query.ps1 -server xxx -port xxx -user xxx -db xxx -q "select xxx"   }
```
server ip is docker ip,or k8s ip

You can batch copy pg_query.ps1 and dependent dlls by k_copyto_ip.ps1

----

[kasini3000](https://gitee.com/chuanjiao10/kasini3000) site mirror:(https://github.com/kasini3000/kasini3000)   
win,linux devops automation batch script framework.(It is similar to Puppet，Ansible，pipeline) 
Open source, free, cross-platform 
English manual: https://gitee.com/chuanjiao10/kasini3000/blob/master/docs/kasini3000_manual.xlsx

[kasini3000_agent_linux](https://gitee.com/chuanjiao10/kasini3000_agent_linux) Shell script,one click install powershell on linux,modify sshd_config for PsRemote.

**Comment by jborean93** [comment]
The problem with `Enter-PSSession` and `Invoke-Command` is that the remote session has not interactive terminal to work with. All communication between the client and the server happen through special PowerShell payloads. So things that are requesting a TTY to interact with the user are going to fail because there is no client end of the TTY to connect to. If you want to get this working through PSRemoting based scenarios then you need to figure out how to do this non-interactively, i.e. don't request `-it` through docker and find ways to run commands from a script rather than user input. If you need user input then don't use PSRemoting, just do `ssh user@host` and work with it like you would in a normal terminal.
